### PR TITLE
Fix TSGFromPropagation

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

TSGFromPropagation was causing a crash when cmsRun was terminated early. This was caused by a pointer being assigned in the init function, not the constructor. This was why the class was inspected. The following changes were made
- all owning pointers were changed to std::unique_ptr
- parameters are not read and assigned in the constructor instead of `init`
- immutable member data are now declared `const`
- unused/unnecessary member variables were removed
- theResetMethod was changed from a `std::string` to an enum.

The latter changed uncovered a bug where two of the string comparisons used with `theResetMethod` were typos: They should have been `"discrete"` but were instead `"discreate"`.

#### PR validation:

The code compiles.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
bugfix

File: RecoMuon/TrackerSeedGenerator/plugins/TSGFromPropagation.cc [hunk 1]
```diff
@@ -375,12 +347,12 @@ void TSGFromPropagation::adjust(FreeTrajectoryState& state) const {
 
 void TSGFromPropagation::adjust(TrajectoryStateOnSurface& state) const {
   //rescale the error
-  if (theResetMethod == "discreate") {
```
Comment (Dr15Jones) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGFromPropagation.cc:378]: This was a bug in the previous code. This should have been `"discrete"`.

File: RecoMuon/TrackerSeedGenerator/plugins/TSGFromPropagation.cc [hunk 2]
```diff
@@ -355,13 +327,13 @@ void TSGFromPropagation::getRescalingFactor(const TrackCand& staMuon) {
 
 void TSGFromPropagation::adjust(FreeTrajectoryState& state) const {
   //rescale the error
-  if (theResetMethod == "discreate") {
```
Comment (Dr15Jones) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGFromPropagation.cc:358]: This was a bug in the previous code. This should have been `"discrete"`.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
bug-fix

**Comment by Dr15Jones** [comment]
type bugfix

**Comment by perrotta** [comment]
+1
- Technical update, plus a bug fix
  - the bug fixed was indeed ineffective, because all settings of  ```ResetMethod``` in the HLT menus are "matrix" (and never "discrete")
- Jenkins tests pass and show no differences in reco code

**Comment by perrotta** [comment]
assign hlt
(perhaps hlt wants to have a look...)

**Comment by Martin-Grunewald** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Pull request merged**